### DISCUSSION TITLE
Support Workspace Access tokens on Bitbucket Cloud

### DISCUSSION THREAD

**Discussion by lupa95** [renovatebot/renovate#21777] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

Hi,

Bitbucket cloud offers _workspace tokens_ as a premium feature, which allow you to assign permissions on a workspace level and are (from their documentation) intended to be used for automation tasks. Using this kind of token seems like a good alternative to bot accounts for me, because the use of bot accounts is sometimes not allowed in corporate environments.

However, I am not sure if it would even be possible to run renovate with this kind of authentication. I don't quite understand, which identity the token would belong to, since it is not bound to any user account.

https://support.atlassian.com/bitbucket-cloud/docs/workspace-access-tokens/

What do you think?

**Comment by viceice** [comment]
simply try it as token on a sample repo. maybe it works out of the box 🤷‍♂️

**Reply by lupa95** [reply]
Thanks for the hint @viceice. I just tried it .... and it sort of works. The bot can create PRs, but it complains about missing permissions for _issues_. Judging from the settings page, there is indeed some difference in assignable permission scopes between app passwords and workspace tokens. While not great, I don't think this is too bad, since the concept of Issues does really not exist on Bitbucket Cloud.

I am a bit more concerned about the PR reviewers feature. The renovate documentation mentions, that you need to add the Bot account to a user group with permissions. This will definitely not work with the workspace token 😄.

```
WARN: Error finding issues (repository=<omitted>)
       "err": {
         "name": "HTTPError",
         "code": "ERR_NON_2XX_3XX_RESPONSE",
         "timings": {
           "start": 1682492598689,
           "socket": 1682492598690,
           "lookup": 1682492598692,
           "connect": 1682492598696,
           "secureConnect": 1682492599018,
           "upload": 1682492599018,
           "response": 1682492599205,
           "end": 1682492599206,
           "phases": {
             "wait": 1,
             "dns": 2,
             "tcp": 4,
             "tls": 322,
             "request": 0,
             "firstByte": 187,
             "download": 1,
             "total": 517
           }
         },
         "message": "Response code 403 (Forbidden)",
         "stack": "HTTPError: Response code 403 (Forbidden)\n    at Request.<anonymous> (/opt/buildpack/tools/renovate/35.58.0/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/35.58.0 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "authorization": "***********",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://api.bitbucket.org/2.0/<omitted>",
           "hostType": "bitbucket",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 403,
           "statusMessage": "Forbidden",
           "body": {
             "type": "error",
             "error": {
               "message": "Your credentials lack one or more required privilege scopes.",
               "detail": {
                 "required": ["issue"],
                 "granted": [
                   "account",
                   "project",
                   "pullrequest:write",
                   "pipeline",
                   "runner"
                 ]
               }
             }
           },
           "headers": {
             "server": "envoy",
             "vary": "Authorization, Origin",
             "content-type": "application/json; charset=utf-8",
             "x-oauth-scopes": "account, project, pullrequest:write, pipeline, runner",
             "x-b3-traceid": "2feb54bf3fd99ff1",
             "x-usage-output-ops": "0",
             "x-used-mesh": "False",
             "x-dc-location": "Micros-3",
             "strict-transport-security": "max-age=31536000; includeSubDomains; preload",
             "date": "Wed, 26 Apr 2023 07:03:20 GMT",
             "x-request-id": "2feb54bf3fd99ff1",
             "x-usage-user-time": "0.036911",
             "x-usage-system-time": "0.004166",
             "x-served-by": "bc76e5e24c2b",
             "x-xss-protection": "1; mode=block",
             "x-envoy-upstream-service-time": "64",
             "x-view-name": "bitbucket.apps.issues.api.v20.issues.RepoIssueListHandler",
             "x-static-version": "68b3568feadf",
             "etag": "\"5fef812c463af18d684e6d17b9d4bdcd\"",
             "x-content-type-options": "nosniff",
             "x-accepted-oauth-scopes": "issue",
             "x-credential-type": "workspace_access_token",
             "x-render-time": "0.05036044120788574",
             "connection": "close",
             "x-trace-id": "2feb54bf3fd99ff1",
             "x-usage-input-ops": "0",
             "x-frame-options": "SAMEORIGIN",
             "x-version": "68b3568feadf",
             "x-request-count": "2266",
             "content-length": "216",
             "x-asap-succeeded": "True"
           },
           "httpVersion": "1.1",
           "retryCount": 0
         }
       }
```

**Reply by viceice** [reply]
so it's probably impossible to use that token. we've a similar issue with gitlab pipeline tokens

**Reply by driverpt** [reply]
If bitbucket doesn't have issues, why does RenovateBot require that permission?

**Reply by rarkins** [reply]
Bitbucket Cloud does have issues support (different from Jira). The better question is whether issues can be skipped for workspace tokens if necessary and the answer would be yes

**Reply by driverpt** [reply]
What about having a settings key to disable `issue:write` permission? Just for the scope of using Workspace/Project Tokens

**Reply by rarkins** [reply]
I don't know what you mean by that. I'm suggesting we just gracefully ignore of issues can't be accessed

**Reply by driverpt** [reply]
I would add a feature flag to not require `issue:write`. Having an implicit behaviour can lead to more issues

**Reply by rarkins** [reply]
I also don't know what you mean by that. Can you express it in Renovate config terminology?

e.g. keep the log warning there by default, users must configure it explicitly if they wish to ignore it?

And would you still think the user should need to configure it even if Renovate could detect which were workspace tokens somehow and log an info message that issues are not available

**Comment by halvorot** [comment]
Any update/solution to this?
I get a 401 when using the workspace access token as RENOVATE_PASSWORD and the related workspace access token email as RENOVATE_USERNAME.
Has anyone got this working?
